druid/src/widget/view_switcher.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget that can dynamically switch between one of many views.
+
+use crate::{
+    BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx,
+    Point, Rect, Size, UpdateCtx, Widget, WidgetPod,
+};
+
+/// A widget that can switch dynamically between one of many views depending
+/// application state.
+pub struct ViewSwitcher<T: Data, U: PartialEq> {
+    child_selector: Box<dyn Fn(&T, &Env) -> U>,
+    child_builder: Box<dyn Fn(&U, &Env) -> Box<dyn Widget<T>>>,
+    active_child: Option<WidgetPod<T, Box<dyn Widget<T>>>>,
+    active_child_id: Option<U>,
+}
+
+impl<T: Data, U: PartialEq> ViewSwitcher<T, U> {
+    /// Create a new view switcher.
+    ///
+    /// The `child_selector` closure is called every time the application data changes.
+    /// If the value it returns is the same as the one it returned during the previous
+    /// data change, nothing happens. If it returns a different value, then the
+    /// `child_builder` closure is called with the new value.
+    ///
+    /// The `child_builder` closure is expected to create a new child widget depending
+    /// on the selector to it.
+    pub fn new(
+        child_selector: impl Fn(&T, &Env) -> U + 'static,
+        child_builder: impl Fn(&U, &Env) -> Box<dyn Widget<T>> + 'static,
+    ) -> Self {
+        Self {
+            child_selector: Box::new(child_selector),
+            child_builder: Box::new(child_builder),
+            active_child: None,
+            active_child_id: None,
+        }
+    }
+}
+
+impl<T: Data, U: PartialEq> Widget<T> for ViewSwitcher<T, U> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        if let Some(ref mut child) = self.active_child {
+            child.event(ctx, event, data, env);
+        }
+    }
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &T, env: &Env) {
+        if let LifeCycle::WidgetAdded = event {
+            let child_id = (self.child_selector)(data, env);
+            self.active_child = Some(WidgetPod::new((self.child_builder)(&child_id, env)));
+            self.active_child_id = Some(child_id);
```

Comment:
Because you're implementing a container widget (a widget that manages one or more other widgets, using a `WidgetPod`) you have certain special considerations, here.

Most importantly, you will actually receive `WidgetAdded` any time one of your children change; the message gets sent to each node in the tree that is an ancestor of whatever widget was actually added; so you should be defensive, and only do this initial setup if we don't have a child, or if we otherwise know that we need to do this setup.

There's another interesting thing happening here: by design, what you _should_ be doing is calling `ctx.children_changed()` after adding your child, and _not_ forwarding the current event to the new child; this will give the framework the chance to send the `WidgetAdded` event, which should always be the first event any new widget receives; however since we're doing this while handling `WidgetAdded` already, this should work as written: it's basically like the child was already here, and it will still receive `WidgetAdded` first and should handle registration etc correctly.

So I think this bit is fine, although in general you should not forward events to newly created widgets, and this definitely needs some more documentation.